### DISCUSSION TITLE
Programmatic use of `fbt-translate`

### DISCUSSION THREAD

**Discussion by alexandernanberg** [facebookincubator/fbt#333] [Q&A]
I'd like to generate translations files on-demand, so translations changes go live without having to deploy the site. E.g.

Translation service (Crowdin) -> Transform translations (`fbt-translate`) -> CDN -> Website

Crowdin have web hooks so the transformation is only run whenever a translation changes. I've got pretty much everything working except the transformation part. AFAIK there is no way of running the transform in plain JS but you must use the script, which isn't convenient  for this particular case. Is there a nice way of getting around this or do I have to inline a bunch of `fbt` internals in my script?

Anyone got pointer or have done something similar?

**Comment by kayhadrin** [comment]
Interesting question timing. I was just thinking of that too: https://github.com/facebook/fbt/issues/334#issuecomment-1023739542

Could you share more context about how you are serving the JS code to browsers?
And how/where the translations are served to the clientside?

With more info about this, we could offer better advice.

**Reply by alexandernanberg** [reply]
So we have a CI job that runs whenever a change is made in our app, running `fbt-manifest`, `fbt-collect` and then uploads the `source_strings.json` file to Crowdin (or whatever service you might use). When a translation is added/changed we trigger a lambda fn that pulls down the `source_strings.json` + translation files from Crowdin and then runs `fbt-transform`. The output is then uploaded to a S3 bucket and served to the apps through a CDN.

<img width="1115" alt="bild" src="https://user-images.githubusercontent.com/8997319/151509938-1422c13a-45c2-41d7-ba0c-381d939c20f3.png">

The old/current setup is the same except without the lambda part, but instead has a daily CI job that pulls down translations, runs `fbt-translate` and creates a PR.
